### PR TITLE
AVX2 optimized audio resampler 

### PR DESCRIPTION
**Pull request by kencooke**
This minimizes the performance impact of using high-quality resampling for all audio content in both Interface and Sandbox.

### CODE REVIEW

**Comment by ZappoMan** [comment]
Test plan: Make sure audio still works. You need to make sure that you're running the audio resampler in this PR.

Here are several tests that will test this new code:

1) run local sandbox with default content (background noises) make sure your sandbox is the sandbox from THIS PR... connect with this client... if you hear sounds... it's running this code and it's working.

2) using any script that injects sounds locally if you hear those sounds... it's running this code... and it's working... examples: collision sounds, ping pong gun.

3) Make sure your'e running dialTone.js with this Client, if this script is running and you connect/disconnect from a domain, and hear the sounds... it's working... it's a pass.

**Comment by kencooke** [comment]
To test, make sure to install both Interface and Sandbox from this build.  Verify that mic audio and environmental sounds are still audible.  This PR does not impact the sound quality, only the performance.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8faaa36913b219e9e235135a9d456ba9e3d592ab available at: https://gist.github.com/2d5338eb9576b8cc4ec31504d0db28ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8024.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8024.dmg

**Comment by ZappoMan** [comment]
hold off on testing until it builds on all platforms.

File: libraries/audio/src/AudioSRC.cpp [hunk 1]
```diff
@@ -911,6 +370,107 @@ int AudioSRC::multirateFilter2(const float* input0, const float* input1, float*
     return outputFrames;
 }
 
+//
+// Detect AVX/AVX2 support
```
Comment (ZappoMan) [inline libraries/audio/src/AudioSRC.cpp]: please move this to `/libraries/shared`

**Comment by hifi-gustavo** [comment]
Build console output for commit: b60b9bb312a4528b620dc384cdfd9b5101f5d117 available at: https://gist.github.com/2d5338eb9576b8cc4ec31504d0db28ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8024.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8024.dmg

**Comment by kencooke** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: b60b9bb312a4528b620dc384cdfd9b5101f5d117 available at: https://gist.github.com/2d5338eb9576b8cc4ec31504d0db28ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8024.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8024.dmg

**Comment by kencooke** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: b60b9bb312a4528b620dc384cdfd9b5101f5d117 available at: https://gist.github.com/2d5338eb9576b8cc4ec31504d0db28ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8024.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8024.dmg

**Comment by kencooke** [comment]
To validate these (now shared) CPU detection functions, I ran this test app via an instruction-set simulator:

```
///////////////////// CPUTest.cpp ///////////////////
#include <stdio.h>
#include "CPUDetect.h"

int main() {
    printf("SSE3  is %s.\n", cpuSupportsSSE3()  ? "enabled" : "disabled");
    printf("SSSE3 is %s.\n", cpuSupportsSSSE3() ? "enabled" : "disabled");
    printf("SSE41 is %s.\n", cpuSupportsSSE41() ? "enabled" : "disabled");
    printf("SSE42 is %s.\n", cpuSupportsSSE42() ? "enabled" : "disabled");
    printf("AVX   is %s.\n", cpuSupportsAVX()   ? "enabled" : "disabled");
    printf("AVX2  is %s.\n", cpuSupportsAVX2()  ? "enabled" : "disabled");
    return 0;
}

```
And got the expected result for each processor:
```
C:\Source\Intel\sde>sde.exe -p4 -- z:CPUTest.exe
SSE3  is disabled.
SSSE3 is disabled.
SSE41 is disabled.
SSE42 is disabled.
AVX   is disabled.
AVX2  is disabled.

C:\Source\Intel\sde>sde.exe -p4p -- z:CPUTest.exe
SSE3  is enabled.
SSSE3 is disabled.
SSE41 is disabled.
SSE42 is disabled.
AVX   is disabled.
AVX2  is disabled.

C:\Source\Intel\sde>sde.exe -mrm -- z:CPUTest.exe
SSE3  is enabled.
SSSE3 is enabled.
SSE41 is disabled.
SSE42 is disabled.
AVX   is disabled.
AVX2  is disabled.

C:\Source\Intel\sde>sde.exe -pnr -- z:CPUTest.exe
SSE3  is enabled.
SSSE3 is enabled.
SSE41 is enabled.
SSE42 is disabled.
AVX   is disabled.
AVX2  is disabled.

C:\Source\Intel\sde>sde.exe -nhm -- z:CPUTest.exe
SSE3  is enabled.
SSSE3 is enabled.
SSE41 is enabled.
SSE42 is enabled.
AVX   is disabled.
AVX2  is disabled.

C:\Source\Intel\sde>sde.exe -wsm -- z:CPUTest.exe
SSE3  is enabled.
SSSE3 is enabled.
SSE41 is enabled.
SSE42 is enabled.
AVX   is disabled.
AVX2  is disabled.

C:\Source\Intel\sde>sde.exe -snb -- z:CPUTest.exe
SSE3  is enabled.
SSSE3 is enabled.
SSE41 is enabled.
SSE42 is enabled.
AVX   is enabled.
AVX2  is disabled.

C:\Source\Intel\sde>sde.exe -ivb -- z:CPUTest.exe
SSE3  is enabled.
SSSE3 is enabled.
SSE41 is enabled.
SSE42 is enabled.
AVX   is enabled.
AVX2  is disabled.

C:\Source\Intel\sde>sde.exe -hsw -- z:CPUTest.exe
SSE3  is enabled.
SSSE3 is enabled.
SSE41 is enabled.
SSE42 is enabled.
AVX   is enabled.
AVX2  is enabled.

C:\Source\Intel\sde>sde.exe -bdw -- z:CPUTest.exe
SSE3  is enabled.
SSSE3 is enabled.
SSE41 is enabled.
SSE42 is enabled.
AVX   is enabled.
AVX2  is enabled.

C:\Source\Intel\sde>sde.exe -skl -- z:CPUTest.exe
SSE3  is enabled.
SSSE3 is enabled.
SSE41 is enabled.
SSE42 is enabled.
AVX   is enabled.
AVX2  is enabled.
```

**Comment by hifi-gustavo** [comment]
Build console output for commit: 27d39eb1b09eb3be43db30985c624669a7d329b0 available at: https://gist.github.com/2d5338eb9576b8cc4ec31504d0db28ad<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8024.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8024.dmg

**Comment by cozza13** [comment]
Using 6th generation. QA Pass
-[06/07 12:39:52] [DEBUG] 	[x] AVX2
-I could hear audio on the sandbox running this PR
-Tested dialtone and that worked. 
- I could hear collision sounds

**Comment by judasshuffle** [comment]
[06/08 09:58:08] [DEBUG] 	[ ] AVX2
damn I'm old inferior again
I could hear myself on echo server
I could hear the fire works on playa
I couldn't hear the ambient noises- that animated robot I'm sure made noises
Processor	Intel(R) Core(TM) i7-2600K CPU @ 3.40GHz, 3701 Mhz, 4 Core(s), 8 Logical Processor(s)
I couldn't hear dial tone on my sandbox or playa but could hear voice in both

**Comment by ZappoMan** [comment]
Interesting judas -- that sounds like a fail.

**Comment by OmegaHeron** [comment]
Testing Linux (Ubuntu 16.04 - QT5.6.0) for stack - Windows 10 for Interface.
Stack shows no issue compiling or running.  Tested on following hardware for stack
Intel i7 Ivybridge
Digital Ocean droplet

Interface tested on i5 Skylake only (has AVX2) - no issues seen (heard)

This was a quick smoke test of audio function only not a full QA, mainly to insure Linux stack is still happy with this.

**Comment by kencooke** [comment]
Judas, if you can reproduce this failure, could you send me the full log file (ken at highfidelity.io)?  All audio sources in Interface (local injector sounds, mic audio, network audio) go through this new resampler so if you hear some but not others, I think you have an unrelated problem.

**Comment by ZappoMan** [comment]
I want to be perfectly clear here--- (hear) --- there's no point in testing stuff in playa... because the audio injectors in playa are not running this code. So when we read this QA report, we are going to ignore any results about sound in playa.

Places that this code path is guaranteed to execute:
- Interface Client calling `AudioInjector::playSound` so that's the dial tone, or collision sounds. It's also in the sandbox if an AC/Agent script calls `Audio.playSound()` which does happen in the "home" content for the background noises.

- Depending on your sound card you **may** see the following message in your logs:
```
Attemping to create a resampler for input format to network format.
```
    - if you see that message in your logs, then your client is also using this code to resample the input from your mic into a format that can be sent to the audio mixer. If you see that message in your log, and you're able to hear **server loop back** -- then that's good this code is running and working. If you can't hear server loopback - it might indicate a problem in this code.

- Depending on your sound card, you **may** see the following message in your logs:
```
Attemping to create a resampler for input format to output format for audio loopback
```
    - if you see that message in your logs, then your client is also using this code to resample the input from your mic into a format that can be played for **local loopback**. If you see this message in your log, and you're able to hear **local loopback** then this code is working. If you see this message and local loopback fails, there may be a problem with this code.

- Depending on your sound card, you **may** see the following message in your logs:
```
Attemping to create a resampler for network format to output format.
```
    - if you see that message in your logs, then your client is also using this code to resample the network data from the audio mixer into a format that can be played to your local speakers. If you see this message in your log, and you're able to hear any server audio then this code is working. If you see this message and server audio fails, there may be a problem with this code.

**Comment by OmegaHeron** [comment]
To be clear - my tests were this PR source built for Linux tested with this PR Interface as provided here.  I also checked this PR Interface against not this PR stack and this PR stack against not this PR Interface.

**Comment by ZappoMan** [comment]
Revision -- if people can hear you (even on a public domain like Playa) and you can hear audio from a domain (even playa) -- then you're running this code... and it's working...

**Pull request merged**